### DISCUSSION TITLE
Why response code 400 during Wemos d1 r1 data sending to firebase???

### DISCUSSION THREAD

**Discussion by hyunseongshin** [community/community#77811] [Programming Help]
### Body

`
#include <ESP8266HTTPClient.h>
#include <ESP8266WiFi.h>
#include <NTPClient.h>
#include <WiFiUdp.h>
#include <DFRobot_DHT11.h>

const char* ssid = "secret";
const char* password = "secret";
const char* firebaseProject = "secret-default-rtdb";
const char* databasePath = "/"; // Firebase 데이터베이스 최상위 경로로 수정

int ledPin = D3; // Connect LED to this pin

WiFiUDP ntpUDP;
NTPClient timeClient(ntpUDP, "pool.ntp.org");

DFRobot_DHT11 DHT;
#define DHT11_PIN D2

// sendDataToFirebase function definition
void sendDataToFirebase(float temperature, float humidity) {
  // Rest of the Firebase transmission code...
  Serial.println("Firebase transmission started");
  timeClient.update();
  String kstTimestamp = convertToKST(timeClient.getEpochTime());
  String data = "{ \"temperature\": " + String(temperature, 2) + ", \"humidity\": " + String(humidity, 2) +
                 ", \"timestamp_kst\": \"" + kstTimestamp + "\" }";

  Serial.println("Data: " + data);

  WiFiClient client;
  HTTPClient http;

  String firebaseURL = "https://" + String(firebaseProject) + ".firebaseio.com" + databasePath;
  http.begin(client, firebaseURL);

  http.addHeader("Content-Type", "application/json");

  // Debug information
  Serial.println("HTTP request started");
  Serial.println("URL: " + firebaseURL);
  Serial.println("Data: " + data);

  int httpResponseCode = http.POST(data);

  if (httpResponseCode > 0) {
    Serial.print("Data sent to Firebase successfully, response code: ");
    Serial.println(httpResponseCode);
  } else {
    Serial.print("Failed to send data to Firebase, response code: ");
    Serial.println(httpResponseCode);

    String response = http.getString();
    Serial.println("Server response: " + response);

    Serial.print("Error: ");
    Serial.println(http.errorToString(httpResponseCode).c_str());
  }

  Serial.println("HTTP request ended");
  http.end();
}

// convertToKST function definition
String convertToKST(time_t utcTime) {
  struct tm *timeinfo;
  timeinfo = gmtime(&utcTime);

  // Convert to KST by adding 9 hours to UTC
  timeinfo->tm_hour += 9;

  char kstBuffer[50];
  strftime(kstBuffer, sizeof(kstBuffer), "%Y-%m-%d %H:%M:%S", timeinfo);

  return String(kstBuffer);
}

void setup() {
  Serial.begin(9600);
  pinMode(ledPin, OUTPUT);
  WiFi.begin(ssid, password);
  
  Serial.println("Setup started");

  while (WiFi.status() != WL_CONNECTED) {
    delay(1000);
    Serial.println("Connecting to WiFi...");
  }

  Serial.println("WiFi connected");
  timeClient.begin();

  while (!timeClient.update()) {
    timeClient.forceUpdate();
  }

  Serial.println("Time synchronization completed");
}

void loop() {
  Serial.println("Loop started");
  DHT.read(DHT11_PIN);
  float temperature = DHT.temperature;
  float humidity = DHT.humidity;

  Serial.println("Temperature: " + String(temperature, 2) + ", Humidity: " + String(humidity, 2));

  if (!isnan(temperature) && !isnan(humidity)) {
    Serial.println("Attempting to send data to Firebase");
    sendDataToFirebase(temperature, humidity);
  }

  delay(5000); // Execute every 5 seconds
}
`

and then it's my serial monitor output

11:21:13.370 -> HTTP request started
11:21:13.370 -> URL: secret/
11:21:13.437 -> Data: { "temperature": 21.00, "humidity": 19.00, "timestamp_kst": "2023-12-04 11:21:12" }
11:21:13.681 -> Data sent to Firebase successfully, response code: 400
11:21:13.717 -> HTTP request ended

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by dakshithadissanayaka** [comment]
The error message "Data sent to Firebase successfully, response code: 400" indicates that the data was transmitted to Firebase, but there was an error processing the request. A 400 error is a Bad Request error, which means that the request sent to the server was invalid or could not be understood.

There are a few potential causes of a 400 error when sending data to Firebase:

* **Invalid JSON data:** The JSON data you are sending to Firebase may be invalid. Make sure that the data is formatted correctly and that all required fields are present.
* **Incorrect database path:** The database path you are specifying may be incorrect. Make sure that the path is correct and that you have permissions to write to that location in the database.
* **Missing authorization header:** You may be missing the authorization header in your request. This header is required to authenticate your request with Firebase.

To troubleshoot this issue, you can try the following:

* **Validate your JSON data:** Use a JSON validator to check that your JSON data is formatted correctly.
* **Verify your database path:** Double-check the database path that you are specifying in your code.
* **Add the authorization header:** Make sure that you are including the authorization header in your request. You can find the authorization header value in the Firebase console.

If you have checked all of these things and you are still getting a 400 error, you may need to contact Firebase support for further assistance.

**Comment by dakshithadissanayaka** [comment]
Or

It looks like the issue is related to the HTTP request sent to Firebase, and the response code `400` suggests that there's a bad request. The Firebase Realtime Database expects the data to be in a specific format, and the issue might be related to the JSON structure.

Here are a few things to check and adjust in your `sendDataToFirebase` function:

1. **Check JSON Syntax:**
   - Ensure that your JSON string is well-formed. In your `data` string, there seems to be an issue with quoting. Instead of double quotes for key-value pairs, you have them as escaped characters. Try updating your `data` string like this:

     ```cpp
     String data = "{ \"temperature\": " + String(temperature, 2) + ", \"humidity\": " + String(humidity, 2) +
                   ", \"timestamp_kst\": \"" + kstTimestamp + "\" }";
     ```

2. **Handle Special Characters:**
   - If the values of temperature, humidity, or timestamp contain special characters, you might need to handle them appropriately. Ensure that there are no characters that could break the JSON format.

3. **Check Firebase URL:**
   - Double-check your `firebaseURL` to make sure it's constructed correctly. The URL should point to your Firebase Realtime Database, and the `databasePath` should be appended to it.

   ```cpp
   String firebaseURL = "https://" + String(firebaseProject) + ".firebaseio.com" + databasePath + ".json";
   ```

4. **Review Firebase Rules:**
   - Make sure that your Firebase Realtime Database rules allow write access. If the rules are too restrictive, it can result in a `400 Bad Request` error.

   ```json
   {
     "rules": {
       ".read": true,
       ".write": true
     }
   }
   ```

Make these adjustments and try running your code again. If the issue persists, you might want to print out the actual error message from Firebase by adding the following lines after `String response = http.getString();`:

```cpp
Serial.println("Server response: " + response);
```

This will provide more details about what Firebase is responding with, which can be helpful in diagnosing the issue.
